cmd/kubeadm/app/discovery/token/token.go

```diff
@@ -131,3 +205,76 @@ func runForEndpointsAndReturnFirst(endpoints []string, fetchKubeConfigFunc func(
 	wg.Wait()
 	return resultingKubeConfig
 }
+
+// certChain is an array of *x509.Certificate with a helper to plug into the tls VerifyPeerCertificate hook
+type certChain []*x509.Certificate
+
+// saveCertificateChain parses and saves the array of DER-encoded certificates into a certChain
+func (c *certChain) saveCertificateChain(rawCerts [][]byte, _ [][]*x509.Certificate) error {
+	// Make sure this wrapper is being used correctly
+	if len(*c) != 0 {
+		return errInsecureClientUsedMoreThanOnce
+	}
+
+	// Do a quick sanity check (this should never happen)
+	if len(rawCerts) < 1 {
+		return errInsecureClientEmptyCertChain
+	}
+
+	// Parse and collect each certificate
+	for _, rawCert := range rawCerts {
+		cert, err := x509.ParseCertificate(rawCert)
+		if err != nil {
+			return err
+		}
+		*c = append(*c, cert)
+	}
+
+	return nil
+}
+
+// withInsecureHTTPClient creates a temporary http.Client, passes it to doRequest(...),
+// which should perform a single HTTP request and return the response. The temporary
+// client will not validate TLS certificates in real time, but instead will capture
+// and return the certificates for post-hoc validation.
+func withInsecureHTTPClient(
```

Comment:
We could extend the normal API client to give more control over the underlying `http.Client`. I think extending it to support this particular case wouldn't be worth it, but a more general change that allowed passing in a custom `http.Client` could be more generally useful.

The thinking with doing this as a "raw" `http.Client` instead of using the normal API client was that this is fairly simple from an HTTP standpoint (single GET request) and that the TLS bits were such an edge case that we probably don't want that complexity in the main API client code.